pkg/modules/tracefunnel/clickhouse_queries.go

```diff
@@ -0,0 +1,790 @@
+    toDateTime64(%[5]d/1e9,9) AS end_ts,
+
+    ('%[6]s','%[7]s') AS step1,
+    ('%[8]s','%[9]s') AS step2,
+    ('%[10]s','%[11]s') AS step3
+
+SELECT
+    count(DISTINCT trace_id) AS total_s1_spans,
+    count(DISTINCT CASE WHEN t1_error = 1 THEN trace_id END) AS total_s1_errored_spans,
+    count(DISTINCT CASE WHEN t2_time > t1_time THEN trace_id END) AS total_s2_spans,
+    count(DISTINCT CASE WHEN t2_time > t1_time AND t2_error = 1 THEN trace_id END) AS total_s2_errored_spans,
+    count(DISTINCT CASE WHEN t2_time > t1_time AND t3_time > t2_time THEN trace_id END) AS total_s3_spans,
+    count(DISTINCT CASE WHEN t2_time > t1_time AND t3_time > t2_time AND t3_error = 1 THEN trace_id END) AS total_s3_errored_spans
+FROM (
+    SELECT
+        trace_id,
+        minIf(timestamp, serviceName = step1.1 AND name = step1.2) AS t1_time,
+        minIf(timestamp, serviceName = step2.1 AND name = step2.2) AS t2_time,
+        minIf(timestamp, serviceName = step3.1 AND name = step3.2) AS t3_time,
+        toUInt8(anyIf(has_error, serviceName = step1.1 AND name = step1.2)) AS t1_error,
+        toUInt8(anyIf(has_error, serviceName = step2.1 AND name = step2.2)) AS t2_error,
+        toUInt8(anyIf(has_error, serviceName = step3.1 AND name = step3.2)) AS t3_error
+    FROM signoz_traces.signoz_index_v3
+    WHERE
+        timestamp BETWEEN start_ts AND end_ts
+        AND (
+            (serviceName = step1.1 AND name = step1.2 AND (contains_error_t1 = 0 OR has_error = true) %[12]s)
+         OR (serviceName = step2.1 AND name = step2.2 AND (contains_error_t2 = 0 OR has_error = true) %[13]s)
+         OR (serviceName = step3.1 AND name = step3.2 AND (contains_error_t3 = 0 OR has_error = true) %[14]s)
+        )
+    GROUP BY trace_id
+    HAVING t1_time > 0 AND t2_time > t1_time AND t3_time > t2_time
+) AS funnel;
+`
+	return fmt.Sprintf(queryTemplate,
+		containsErrorT1,
+		containsErrorT2,
+		containsErrorT3,
+		startTs,
+		endTs,
+		serviceNameT1,
+		spanNameT1,
+		serviceNameT2,
+		spanNameT2,
+		serviceNameT3,
+		spanNameT3,
+		clauseStep1,
+		clauseStep2,
+		clauseStep3,
+	)
+}
+
+func BuildTwoStepFunnelTopSlowTracesQuery(
+	containsErrorT1 int,
+	containsErrorT2 int,
+	startTs int64,
+	endTs int64,
+	serviceNameT1 string,
+	spanNameT1 string,
+	serviceNameT2 string,
+	spanNameT2 string,
+	clauseStep1 string,
+	clauseStep2 string,
+) string {
+	queryTemplate := `
+WITH
+    %[1]d AS contains_error_t1,
+    %[2]d AS contains_error_t2,
+    toDateTime64(%[3]d/1e9, 9) AS start_ts,
+    toDateTime64(%[4]d/1e9, 9) AS end_ts,
+
+    ('%[5]s','%[6]s') AS step1,
+    ('%[7]s','%[8]s') AS step2
+
+SELECT
+    trace_id,
+    (toUnixTimestamp64Nano(t2_time) - toUnixTimestamp64Nano(t1_time)) / 1e6 AS duration_ms,
+    span_count
+FROM (
+    SELECT
+        trace_id,
+        minIf(timestamp, serviceName = step1.1 AND name = step1.2) AS t1_time,
+        minIf(timestamp, serviceName = step2.1 AND name = step2.2) AS t2_time,
+        count() AS span_count
```

Comment:
This won't give correct count of spans in the traces as we filtered the spans in the trace before this

Reply:
removed from the feature